flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,340 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
 } // namespace Fortran
 
 namespace {
+namespace looputils {
+
+/// Stores info needed about the induction/iteration variable for each `do
+/// concurrent` in a loop nest. This includes:
+/// * the operation allocating memory for iteration variable,
+/// * the operation(s) updating the iteration variable with the current
+///   iteration number.
+struct InductionVariableInfo {
+  mlir::Operation *iterVarMemDef;
+  llvm::SetVector<mlir::Operation *> indVarUpdateOps;
+};
+
+using LoopNestToIndVarMap =
+    llvm::MapVector<fir::DoLoopOp, InductionVariableInfo>;
+
+/// Given an operation `op`, this returns true if `op`'s operand is ultimately
+/// the loop's induction variable. Detecting this helps finding the live-in
```

Comment:
Why just convert and store operations? Shouldn't this trigger on anything that depends (directly or indirectly) on the induction variables. Depending on a particular form of the body should be avoided.

Reply:
Re-implemented more generically.